inference-engine/src/transformations/include/transformations/common_optimizations/common_optimizations_tbl.hpp

```diff
@@ -18,6 +18,7 @@
 NGRAPH_PASS(InitNodeInfo, ::ngraph::pass)
 NGRAPH_PASS(RemoveFilteringBoxesBySize, ::ngraph::pass) // Resolves dynamism (replaces NonZero), CF needed
 NGRAPH_PASS(ConstantFolding, ::ngraph::pass)
+NGRAPH_PASS(UpgradeNMS3ToNMS4, ::ngraph::pass) // replaces v3::NMS with v4::NMS(always dyn output shape) if function has opset3::NonZero operation
```

Comment:
First of all -- I do not track dynamic branches here at all since the purpose of this pass is to introduce new dynamism inside the model. It is a complicated question -- I would like to take this discussion offline.